src/array.jl

```diff
@@ -54,7 +58,10 @@ function unsafe_free!(a::CuArray)
     ptr = pointer(a)
     if isvalid(ptr.ctx)
         @trace("Finalizing CuArray object at $(Base.pointer_from_objref(a))")
-        Mem.free(ptr)
+        if Mem.refcount(ptr) == 1
```

Comment:
>  could also be used without the creation of Julia objects + finalizer
Why wouldn't you be able to do that? Putting the refcounting logic in `memory.jl` would only imply that `free` will only be called once after doing a `malloc`, it wouldn't change the externally-facing API, ie. you don't need finalizers.

In addition, it would make it easy to re-use the refcounting logic, for when people want to define their own array types.